// @figma/code-connect/dist/connect/migration_batch_helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AUTO_BATCH_THRESHOLD = void 0;
exports.getBatchMigrationGroups = getBatchMigrationGroups;
exports.buildBatchMigration = buildBatchMigration;
exports.writeBatchTemplateFiles = writeBatchTemplateFiles;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const migration_helpers_1 = require("./migration_helpers");
exports.AUTO_BATCH_THRESHOLD = 10;
function uniqueValues(values) {
    return new Set(values);
}
function getBatchMigrationGroups(docs, { batchAll = false, disabled = false, threshold = exports.AUTO_BATCH_THRESHOLD, } = {}) {
    if (disabled) {
        return new Map();
    }
    const docsBySourcePath = new Map();
    for (const doc of docs) {
        if (!doc._codeConnectFilePath) {
            continue;
        }
        const sourcePath = path_1.default.resolve(doc._codeConnectFilePath);
        docsBySourcePath.set(sourcePath, [...(docsBySourcePath.get(sourcePath) || []), doc]);
    }
    const batchGroups = new Map();
    for (const [sourcePath, sourceDocs] of docsBySourcePath) {
        if (batchAll || sourceDocs.length >= threshold) {
            batchGroups.set(sourcePath, sourceDocs);
        }
    }
    return batchGroups;
}
function allEqual(values) {
    return values.every((value) => value === values[0]);
}
function isIdentifier(value) {
    return !!value && /^[A-Za-z_$][\w$]*$/.test(value);
}
function escapeRegExp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
function applyReplacements(body, replacements) {
    return replacements
        .sort((a, b) => b.start - a.start)
        .reduce((result, replacement) => {
        return result.slice(0, replacement.start) + replacement.value + result.slice(replacement.end);
    }, body);
}
function removeUnusedSimpleConsts(body) {
    let previous;
    let next = body;
    const simpleConstRegex = /^const\s+([A-Za-z_$][\w$]*)\s*=\s*(?:"[^"\n]*"|'[^'\n]*'|figma\.helpers\.react\.identifier\(["'][^"'\n]+["']\)|[A-Za-z_$][\w$]*|true|false|-?\d+(?:\.\d+)?)\s*\n/gm;
    do {
        previous = next;
        next = next.replace(simpleConstRegex, (match, name, offset) => {
            const rest = next.slice(offset + match.length);
            const isReferenced = new RegExp(`\\b${escapeRegExp(name)}\\b`).test(rest);
            return isReferenced ? match : '';
        });
    } while (next !== previous);
    return next;
}
function deriveMetadata(docs) {
    const sources = docs.map((doc) => doc.source || undefined);
    const components = docs.map((doc) => doc.component || undefined);
    const entries = docs.map((doc) => ({ url: doc.figmaNode }));
    const comments = [];
    const metadataFields = [
        { name: 'source', values: sources },
        { name: 'component', values: components },
    ];
    for (const field of metadataFields) {
        const hasAny = field.values.some(Boolean);
        const hasAll = field.values.every(Boolean);
        const unique = uniqueValues(field.values);
        if (hasAny && !hasAll) {
            return {
                ok: false,
                reason: `Cannot batch: some docs have ${field.name} metadata and others do not`,
            };
        }
        if (!hasAny) {
            continue;
        }
        if (unique.size === 1) {
            comments.push(`// ${field.name}=${field.values[0]}`);
        }
        else {
            docs.forEach((doc, index) => {
                const value = doc[field.name];
                if (value) {
                    entries[index][field.name] = value;
                }
            });
        }
    }
    return { ok: true, comments, entries };
}
function parseImport(importString) {
    const named = importString.match(/^import\s+\{\s*([A-Za-z_$][\w$]*)\s*\}\s+from\s+["']([^"']+)["']\s*;?$/);
    if (named) {
        return { kind: 'named', symbol: named[1], importPath: named[2], raw: importString };
    }
    const defaultImport = importString.match(/^import\s+([A-Za-z_$][\w$]*)\s+from\s+["']([^"']+)["']\s*;?$/);
    if (defaultImport) {
        return {
            kind: 'default',
            symbol: defaultImport[1],
            importPath: defaultImport[2],
            raw: importString,
        };
    }
    return null;
}
function tryIdSwap(candidates) {
    const matches = candidates.map((candidate) => candidate.body.match(/\bid:\s*(['"])([^'"\\]+)\1/));
    if (matches.some((match) => !match)) {
        return candidates;
    }
    const ids = matches.map((match) => match[2]);
    if (uniqueValues(ids).size <= 1) {
        return candidates;
    }
    return candidates.map((candidate, index) => ({
        ...candidate,
        body: candidate.body.replace(matches[index][0], 'id: figma.batch.id'),
        params: { ...candidate.params, id: ids[index] },
    }));
}
function tryImportSwap(candidates) {
    const importLists = candidates.map((candidate) => candidate.doc.templateData?.imports || []);
    if (importLists.some((imports) => imports.length !== 1)) {
        return candidates;
    }
    const importStrings = importLists.map((imports) => imports[0]);
    if (uniqueValues(importStrings).size <= 1) {
        return candidates;
    }
    const parsed = importStrings.map(parseImport);
    if (parsed.some((info) => !info)) {
        return candidates;
    }
    const imports = parsed;
    if (!allEqual(imports.map((info) => info.kind))) {
        return candidates;
    }
    const symbolVaries = uniqueValues(imports.map((info) => info.symbol)).size > 1;
    const pathVaries = uniqueValues(imports.map((info) => info.importPath)).size > 1;
    if (!symbolVaries && !pathVaries) {
        return candidates;
    }
    return candidates.map((candidate, index) => {
        const info = imports[index];
        const symbol = symbolVaries ? '${figma.batch.componentName}' : info.symbol;
        const importPath = pathVaries ? '${figma.batch.importPath}' : info.importPath;
        const replacement = info.kind === 'named'
            ? `\`import { ${symbol} } from "${importPath}"\``
            : `\`import ${symbol} from "${importPath}"\``;
        const encodedImport = JSON.stringify(info.raw);
        if (!candidate.body.includes(encodedImport)) {
            return candidate;
        }
        return {
            ...candidate,
            body: candidate.body.replace(encodedImport, replacement),
            params: {
                ...candidate.params,
                ...(symbolVaries ? { componentName: info.symbol } : {}),
                ...(pathVaries ? { importPath: info.importPath } : {}),
            },
        };
    });
}
function templateLiteralTextRanges(content, contentStart) {
    const ranges = [];
    let textStart = 0;
    let i = 0;
    while (i < content.length) {
        if (content[i] === '$' && content[i + 1] === '{') {
            if (textStart < i) {
                ranges.push({
                    start: contentStart + textStart,
                    end: contentStart + i,
                    text: content.slice(textStart, i),
                });
            }
            i += 2;
            let depth = 1;
            while (i < content.length && depth > 0) {
                if (content[i] === '{')
                    depth++;
                if (content[i] === '}')
                    depth--;
                i++;
            }
            textStart = i;
            continue;
        }
        i++;
    }
    if (textStart < content.length) {
        ranges.push({
            start: contentStart + textStart,
            end: contentStart + content.length,
            text: content.slice(textStart),
        });
    }
    return ranges;
}
function figmaCodeTextRanges(body) {
    const ranges = [];
    const regex = /figma\.code`([\s\S]*?)`/g;
    let match;
    while ((match = regex.exec(body))) {
        const contentStart = match.index + 'figma.code`'.length;
        ranges.push(...templateLiteralTextRanges(match[1], contentStart));
    }
    return ranges;
}
function replaceIdentifierInFigmaCode(body, identifier, replacement) {
    const ranges = figmaCodeTextRanges(body);
    const identifierRegex = new RegExp(`(^|[^A-Za-z0-9_$])${escapeRegExp(identifier)}(?=$|[^A-Za-z0-9_$])`, 'g');
    const replacements = [];
    for (const range of ranges) {
        let match;
        while ((match = identifierRegex.exec(range.text))) {
            const prefixLength = match[1].length;
            const start = range.start + match.index + prefixLength;
            replacements.push({
                start,
                end: start + identifier.length,
                value: replacement,
            });
        }
    }
    return applyReplacements(body, replacements);
}
function tryComponentSymbolSwap(candidates) {
    let withNames = candidates;
    if (withNames.some((candidate) => !candidate.params.componentName)) {
        const ids = withNames.map((candidate) => candidate.params.id);
        if (ids.every(isIdentifier) && uniqueValues(ids).size > 1) {
            withNames = withNames.map((candidate) => ({
                ...candidate,
                params: { ...candidate.params, componentName: candidate.params.id },
            }));
        }
    }
    if (withNames.some((candidate) => !candidate.params.componentName)) {
        return candidates;
    }
    const names = withNames.map((candidate) => candidate.params.componentName);
    if (uniqueValues(names).size <= 1) {
        return candidates;
    }
    return withNames.map((candidate, index) => ({
        ...candidate,
        body: replaceIdentifierInFigmaCode(candidate.body, names[index], '${figma.batch.componentName}'),
    }));
}
function collectNamedScalars(body, kind) {
    const ranges = figmaCodeTextRanges(body);
    const regex = kind === 'string'
        ? /([A-Za-z_$][\w$]*)\s*(=|:)\s*(['"])([^'"\\\n]+)\3/g
        : /([A-Za-z_$][\w$]*)\s*(=|:)\s*(?:\{\s*)?(true|false|-?\d+(?:\.\d+)?)(?:\s*\})?(?=$|[^A-Za-z0-9_$.])/g;
    const matches = [];
    for (const range of ranges) {
        let match;
        while ((match = regex.exec(range.text))) {
            const value = kind === 'string' ? match[4] : match[3];
            const valueOffset = match[0].lastIndexOf(value);
            matches.push({
                key: match[1],
                operator: match[2],
                value,
                valueStart: range.start + match.index + valueOffset,
                valueEnd: range.start + match.index + valueOffset + value.length,
            });
        }
    }
    return matches;
}
function parseScalarValue(value, kind) {
    if (kind === 'string') {
        return value;
    }
    if (value === 'true') {
        return true;
    }
    if (value === 'false') {
        return false;
    }
    return Number(value);
}
function tryNamedScalarSwap(candidates, kind) {
    const matchesByCandidate = candidates.map((candidate) => collectNamedScalars(candidate.body, kind));
    const matchCount = matchesByCandidate[0]?.length || 0;
    if (matchCount === 0 || matchesByCandidate.some((matches) => matches.length !== matchCount)) {
        return candidates;
    }
    const replacementsByCandidate = candidates.map(() => []);
    const fieldNames = new Set();
    const nextCandidates = candidates.map((candidate) => ({
        ...candidate,
        params: { ...candidate.params },
    }));
    for (let matchIndex = 0; matchIndex < matchCount; matchIndex++) {
        const matches = matchesByCandidate.map((matches) => matches[matchIndex]);
        const first = matches[0];
        if (matches.some((match) => match.key !== first.key || match.operator !== first.operator) ||
            uniqueValues(matches.map((match) => match.value)).size <= 1) {
            continue;
        }
        if (fieldNames.has(first.key) || first.key in nextCandidates[0].params) {
            return candidates;
        }
        fieldNames.add(first.key);
        matches.forEach((match, candidateIndex) => {
            replacementsByCandidate[candidateIndex].push({
                start: match.valueStart,
                end: match.valueEnd,
                value: `\${figma.batch.${first.key}}`,
            });
            nextCandidates[candidateIndex].params[first.key] = parseScalarValue(match.value, kind);
        });
    }
    if (fieldNames.size === 0) {
        return candidates;
    }
    return nextCandidates.map((candidate, index) => ({
        ...candidate,
        body: applyReplacements(candidate.body, replacementsByCandidate[index]),
    }));
}
function buildBatchMigration(docs, { includeProps = false, useTypeScript = true, templateFile = './template.figma.batch.ts', } = {}) {
    if (docs.length < 2) {
        return { ok: false, reason: 'Cannot batch: at least two docs are required' };
    }
    if (docs.some((doc) => doc.variant && Object.keys(doc.variant).length > 0)) {
        return { ok: false, reason: 'Cannot batch: variant Code Connect docs are not supported yet' };
    }
    if (docs.some((doc) => !doc.template)) {
        return { ok: false, reason: 'Cannot batch: every doc must have a generated template' };
    }
    const metadata = deriveMetadata(docs);
    if (!metadata.ok) {
        return metadata;
    }
    let candidates = docs.map((doc, index) => ({
        doc,
        body: removeUnusedSimpleConsts((0, migration_helpers_1.prepareMigratedTemplateBody)(doc, includeProps, useTypeScript)),
        params: metadata.entries[index],
    }));
    candidates = tryIdSwap(candidates);
    candidates = tryImportSwap(candidates);
    candidates = tryComponentSymbolSwap(candidates);
    candidates = tryNamedScalarSwap(candidates, 'string');
    candidates = tryNamedScalarSwap(candidates, 'numberBoolean');
    if (!allEqual(candidates.map((candidate) => candidate.body))) {
        return {
            ok: false,
            reason: 'Cannot batch: templates did not reduce to one compatible shape',
        };
    }
    const templateBody = (0, migration_helpers_1.formatTemplate)(candidates[0].body);
    const comments = metadata.comments.length > 0 ? `${metadata.comments.join('\n')}\n\n` : '';
    return {
        ok: true,
        template: `${comments}${templateBody}`,
        batchJson: {
            templateFile,
            components: candidates.map((candidate) => candidate.params),
        },
    };
}
function getSourceBasename(localSourcePath, docs) {
    if (!localSourcePath) {
        return (0, migration_helpers_1.getFilenameFromComponentName)(docs[0]?.component || 'batch');
    }
    const sourceBasename = path_1.default.basename(localSourcePath);
    const codeConnectPattern = /\.(figma|figmadoc)(\.[^.]+)+$/;
    return codeConnectPattern.test(sourceBasename)
        ? sourceBasename.replace(codeConnectPattern, '')
        : path_1.default.basename(localSourcePath, path_1.default.extname(localSourcePath));
}
function determineBatchOutputPaths(basename, outputDir, baseDir, localSourcePath, useTypeScript, filePathsCreated) {
    const dir = outputDir || (localSourcePath ? path_1.default.dirname(localSourcePath) : baseDir);
    const templateSuffix = useTypeScript ? '.figma.batch.ts' : '.figma.batch.js';
    const getPaths = (name) => ({
        templatePath: path_1.default.join(dir, `${name}${templateSuffix}`),
        batchPath: path_1.default.join(dir, `${name}.figma.batch.json`),
    });
    const basePaths = getPaths(basename);
    if ((fs_1.default.existsSync(basePaths.templatePath) && !filePathsCreated?.has(basePaths.templatePath)) ||
        (fs_1.default.existsSync(basePaths.batchPath) && !filePathsCreated?.has(basePaths.batchPath))) {
        return { skipped: true, reason: 'batch output file already exists' };
    }
    let paths = basePaths;
    let counter = 1;
    while (fs_1.default.existsSync(paths.templatePath) ||
        fs_1.default.existsSync(paths.batchPath) ||
        filePathsCreated?.has(paths.templatePath) ||
        filePathsCreated?.has(paths.batchPath)) {
        paths = getPaths(`${basename}_${counter}`);
        counter++;
    }
    return { skipped: false, ...paths };
}
function writeBatchTemplateFiles(docs, outputDir, baseDir, { localSourcePath, filePathsCreated, includeProps = false, useTypeScript = true, } = {}) {
    const basename = getSourceBasename(localSourcePath, docs);
    const paths = determineBatchOutputPaths(basename, outputDir, baseDir, localSourcePath, useTypeScript, filePathsCreated);
    if (paths.skipped) {
        return { skipped: true, reason: paths.reason, componentCount: docs.length };
    }
    const built = buildBatchMigration(docs, {
        includeProps,
        useTypeScript,
        templateFile: `./${path_1.default.basename(paths.templatePath)}`,
    });
    if (!built.ok) {
        return { skipped: true, reason: built.reason, componentCount: docs.length };
    }
    fs_1.default.mkdirSync(path_1.default.dirname(paths.templatePath), { recursive: true });
    fs_1.default.writeFileSync(paths.templatePath, built.template, 'utf-8');
    fs_1.default.writeFileSync(paths.batchPath, `${JSON.stringify(built.batchJson, null, 2)}\n`, 'utf-8');
    filePathsCreated?.add(paths.templatePath);
    filePathsCreated?.add(paths.batchPath);
    return {
        skipped: false,
        templatePath: paths.templatePath,
        batchPath: paths.batchPath,
        componentCount: docs.length,
    };
}
